### DISCUSSION TITLE
Calculate apparentTemperature and use it for feelsLikeTemperature

### DISCUSSION THREAD

**Discussion by chunshek** [breezy-weather/breezy-weather#1085] [Ideas]
### Describe your idea

I would like to suggest adopting the [Steadman Apparent Temperature formula](http://www.bom.gov.au/info/thermal_stress/#atapproximation) to calculate the `apparentTemperature` field, and using that to fill in `feelsLikeTemperature`.

Currently, Breezy Weather calculates the `wetBulbTemperature` if it has the `temperature` and `relativeHumidity` for a forecast time, using a [regression formula developed by Roland Stull](https://journals.ametsoc.org/view/journals/apme/50/11/jamc-d-11-0143.1.xml). `wetBulbTemperature` is used to fill in `feelsLikeTemperature` in the absence of `realFeelTemperature`, `realFeelShaderTemperature` or `apparentTemperature`.

The [Wet Bulb Temperature](https://en.wikipedia.org/wiki/Wet-bulb_temperature) (not the same thing as the Wet Bulb Globe Temperature) is the reading from a thermometer wrapped in a water-soaked cloth. At relative humidity below 100%, it is always lower than the ambient temperature due to evaporative cooling. It measures the lowest temperature that can be achieved to a wet surface due to evaporation.

Using the `wetBulbTemperature` as a stand-in for `feelsLikeTemperature` underestimates the danger of heat stress, because it is not directly comparable to what the ambient temperature would "feel like" to the human body under normal conditions: we aren't walking around wrapped in a wet towel at all times.

The [Wet Bulb Globe Temperature](https://en.wikipedia.org/wiki/Wet-bulb_globe_temperature) (WBGT) is a combination of three different temperature measurements: 70% Wet Bulb Temperature, 20% Globe Thermometer Temperature (taken with a "black globe thermometer" to measure radiant heat from the sun) and 10% Ambient Temperature. The WBGT is a good measurement for assessing health risk due to heat stress, but is still not what the body "feels like".

### Other details

The Steadman Apparent Temperature formula was invented by Robert G. Steadman, and is used by the Bureau of Meteorology of Australia in its weather reports. It covers both the perceived effects for hot and cold weather conditions. The hot weather version was adopted by the National Weather Service in the US for calculating Heat Index. (Source: [http://www.bom.gov.au/info/thermal_stress/](http://www.bom.gov.au/info/thermal_stress/))

The Apparent Temperature formula is as follows:

> _AT_ = _Ta_ + 0.33 × _e_ − 0.70 × _ws_ − 4.00

where _Ta_ is the ambient temperature in °C
_e_ is the water vapor pressure in hPa
_ws_ is the wind speed in m/s

The water vapor pressure (_e_) is calculated as:

> _e_ = _rh_ / 100 × 6.105 × exp(17.27 × _Ta_ / (237.7 + _Ta_))

where _rh_ is the relative humidity in %

I applied the Steadman AT formula (_at_) and the Stull WBT formula (_wb_) against some forecast data from CWA for Pingtung City, Pingtung County, Taiwan (22.6627972, 120.4915507). The Steadman formula (_at_) largely corresponds to the official CWA forecast for apparent temperatures (AT), whereas the Stull formula (_wbt_) largely corresponds to the official CWA forecast for comfort index (CI). Notice that the difference between the two can sometimes be as much as 10°C, e.g. at `2024-06-02T15:00:00` (Steadman 40°C vs. Stull 30°C) and `2024-06-05T15:00:00` (Steadman 38°C vs. Stull 28°C).

| Time                | T  | RH | WS | AT | CI | _at_ | _wbt_
|---------------------|----|----|----|----|----|------|-------
| 2024-06-02T06:00:00 | 27 | 89 |  2 | 32 | 26 |   32 |    26 
| 2024-06-02T09:00:00 | 30 | 85 |  4 | 36 | 29 |   35 |    28 
| 2024-06-02T12:00:00 | 32 | 80 |  5 | 37 | 30 |   37 |    29 
| 2024-06-02T15:00:00 | 32 | 84 |  2 | 37 | 31 |   40 |    30 
| 2024-06-02T18:00:00 | 30 | 84 |  1 | 36 | 29 |   37 |    28 
| 2024-06-02T21:00:00 | 28 | 86 |  1 | 33 | 27 |   34 |    26 
| 2024-06-03T00:00:00 | 27 | 88 |  2 | 32 | 26 |   32 |    25 
| 2024-06-03T03:00:00 | 27 | 91 |  1 | 31 | 26 |   33 |    26 
| 2024-06-03T06:00:00 | 26 | 90 |  2 | 31 | 25 |   31 |    25 
| 2024-06-03T09:00:00 | 29 | 76 |  2 | 34 | 27 |   34 |    26 
| 2024-06-03T12:00:00 | 31 | 67 |  3 | 35 | 28 |   35 |    26 
| 2024-06-03T15:00:00 | 31 | 71 |  3 | 35 | 28 |   35 |    27 
| 2024-06-03T18:00:00 | 29 | 76 |  3 | 33 | 27 |   33 |    26 
| 2024-06-03T21:00:00 | 28 | 82 |  2 | 31 | 26 |   33 |    26 
| 2024-06-04T00:00:00 | 27 | 87 |  1 | 31 | 26 |   33 |    25 
| 2024-06-04T03:00:00 | 26 | 91 |  1 | 31 | 26 |   31 |    25 
| 2024-06-04T06:00:00 | 26 | 91 |  1 | 30 | 25 |   31 |    25 
| 2024-06-04T09:00:00 | 29 | 76 |  1 | 34 | 27 |   34 |    26 
| 2024-06-04T12:00:00 | 32 | 63 |  2 | 36 | 28 |   36 |    26 
| 2024-06-04T15:00:00 | 32 | 71 |  2 | 36 | 29 |   38 |    28 
| 2024-06-04T18:00:00 | 30 | 77 |  1 | 34 | 28 |   36 |    27 
| 2024-06-04T21:00:00 | 28 | 83 |  1 | 32 | 26 |   34 |    26 
| 2024-06-05T00:00:00 | 27 | 88 |  0 | 32 | 26 |   33 |    25 
| 2024-06-05T03:00:00 | 26 | 90 |  1 | 31 | 25 |   31 |    25 
| 2024-06-05T06:00:00 | 25 | 90 |  1 | 30 | 25 |   30 |    24 
| 2024-06-05T09:00:00 | 29 | 83 |  2 | 35 | 28 |   35 |    27 
| 2024-06-05T12:00:00 | 31 | 72 |  2 | 36 | 29 |   36 |    27 
| 2024-06-05T15:00:00 | 31 | 68 |  2 | 36 | 28 |   36 |    26 
| 2024-06-05T18:00:00 | 30 | 77 |  1 | 34 | 28 |   36 |    27 
| 2024-06-05T21:00:00 | 28 | 85 |  1 | 33 | 27 |   34 |    26 
| 2024-06-06T00:00:00 | 27 | 89 |  1 | 32 | 26 |   33 |    26 
| 2024-06-06T03:00:00 | 26 | 90 |  1 | 31 | 25 |   31 |    25

### Acknowledgements

- [X] I have searched the [existing issues](https://github.com/breezy-weather/breezy-weather/issues) and [existing discussions](https://github.com/breezy-weather/breezy-weather/discussions/categories/ideas) and this is a new ticket, **NOT** a duplicate or related to another open OR closed issue or discussion.
- [X] The idea I’m suggesting is not in the [list of features that will not be implemented](https://github.com/breezy-weather/breezy-weather/blob/main/README.md#features-that-will-not-be-implemented).
- [X] I have updated the app to **[version 5.2.3 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)**. Even if I’m getting updates from F-Droid default repo, I don’t need to uninstall/reinstall to update.
- [X] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance). I can write in my native language under the English text, if necessary.
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by papjul** [comment]
> Using the wetBulbTemperature as a stand-in for feelsLikeTemperature underestimates the danger of heat stress

I was more interested by the formula with sun radiation (a data we do not store *yet*), but I couldn't find it back then. Maybe you have it?

The simplified formula still depends on more factors that may not be available with the source, such as pressure, so you could still have a fallback to wet bulb if implemented. But in all cases, wet bulb is always last to be displayed as feels like temperature after all feels like temperature we can have have been tried.

**Reply by chunshek** [reply]
The version with solar radiation is as follows:

> _AT_ = _Ta_ + 0.348 × _e_ − 0.70 × _ws_ + 0.70 × _Q_ / (_ws_ + 10) − 4.25

where _Ta_ is the ambient temperature in °C
_e_ is the water vapor pressure in hPa
_ws_ is the wind speed in m/s
_Q_ is the solar radiation per unit area of body surface in w/m²

The Bureau of Meteorology in Australia officially uses the non-radiation version in its forecast. It appears that the CWA in Taiwan uses the non-radiation version as well. I guess very few agencies would provide the solar radiation per unit area in their forecast data! (I don't know any that do.)

By the way, the "water vapor pressure" (_e_) is actually a function of the relative humidity and ambient temperature. It has nothing to do with air pressure:

> _e_ = _rh_ / 100 × 6.105 × exp(17.27 × _Ta_ / (237.7 + _Ta_))

So, as long as we have (1) ambient temperature, (2) relative humidity, (3) wind speed – which many, though not all, forecasts provide – we can combine _e_ into the main formula as follows:

> **Non-radiation formula**
> _AT_ = _Ta_ + 0.33 × _rh_ / 100 × 6.105 × exp(17.27 × _Ta_ / (237.7 + _Ta_)) − 0.70 × _ws_ − 4.00

> **Solar radiation formula**
> _AT_ = _Ta_ + 0.348 × _rh_ / 100 × 6.105 × exp(17.27 × _Ta_ / (237.7 + _Ta_)) − 0.70 × _ws_ + 0.70 × _Q_ / (_ws_ + 10) − 4.25

I guess my point was that there are standard formulae for calculating "apparent temperature" that Breezy can already use. The "wet bulb temperature" is not quite how our body feels like, unless we walk around in a wet towel all day. Living in a hot country, seeing "wet bulb temperature" as a stand-in for "feels like temperature" just seems very off. It definitely feels like 39°C today where I live, rather than the 31°C that the wet bulb suggests! 🥵

(I have a [small branch](https://github.com/chunshek/breezy-weather/commit/f9caf926cafdde674899ba4064e0ffc3f2037579) with the above formula added, but you can probably apply the change yourself unless you want me to push a PR.)

**Reply by papjul** [reply]
> (I have a [small branch](https://github.com/chunshek/breezy-weather/commit/f9caf926cafdde674899ba4064e0ffc3f2037579) with the above formula added, but you can probably apply the change yourself unless you want me to push a PR.)

Done, merged

**Reply by papjul** [reply]
I wanted to use the solar radiation version in priority (and fallback to the other one) as I thought it would be more useful, but if most -if not all- sources don't use it, it's just easier.
